### ISSUE TITLE
TODO: Add usage instructions of SPM and swiftc

### ISSUE THREAD

**Issue by stevapple** [compnerd/swift-build#284]

**Comment by stevapple** [comment]
Willing to help, but I’m a little busy now😢Leave it here and, if anyone wants to update this, just go ahead.

**Comment by compnerd** [comment]
Happy to get improvements on the documentation!

Using `swiftc` requires a few flags as documented in the [Getting Started](https://github.com/compnerd/swift-build/blob/master/docs/GettingStartedWindows.md#minimal-hello-world-cmake).

```cmd
set SDKROOT=%SystemDrive%/Library/Developer/Platforms/Windows.platform/Developer/SDKs/Windows.sdk
set SWIFTFLAGS=-sdk %SDKROOT% -I %SDKROOT%/usr/lib/swift -L %SDKROOT%/usr/lib/swift/windows
swiftc %SWIFTFLAGS% -libc MD -O -emit-executable -o executable.exe executable.swift
```

swift-package-manager is still under active development, and it is not fully ready for general usage.  It is possible to use, but it comes with the assumption that things will change without notice and expectations of the user will change without notice.  For using swift-package-manager, assuming you are using an installed image, something like this _should_ work:

```cmd
set SDKROOT=%SystemDrive%/Library/Developer/Platforms/Windows.platform/Developer/SDKs/Windows.sdk
swift build -Xmanifest -resource-dir -Xmanifest %SDKROOT%\usr\lib\swift -Xmanifest -L%SDKROOT%\usr\lib\swift\windows -Xmanifest -libc -Xmanifest MD -Xswiftc -resource-dir -Xswiftc %SDKROOT%\usr\lib\swift -Xswiftc -L%SDKROOT%\usr\lib\swift\windows -Xswiftc -libc -Xswiftc MD
```

**Comment by compnerd** [comment]
I think that we should probably split this up into two separate disjoint things.  The s-p-m commands are about to change for windows.  It will require the latest builds from azure, and the installer.  This is because there are expectations around environment variables and existence of plists and directory structure.  It takes into account the split toolchain/SDK setup and will actually be expected to be used as `swift build` without any additional flags after the toolchain has been installed by the installer.

**Comment by compnerd** [comment]
SPM usage is now just `swift build` or `swift test`.  The `swiftc` invocations still require additional flags (intentionally).

**Comment by RUSshy** [comment]
Hello,

Are the changes in the new toolchain available here: https://swift.org/download/#releases?

I tried using that one, and i still get: ``error: unable to invoke subcommand: C:\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin\swift-build ()``

**Comment by stevapple** [comment]
@RUSshy Nope, `swift build` requires SPM, which is not included in the 5.3 release for Windows. If you use that toolchain, you may need to use CMake.

**Comment by compnerd** [comment]
@RUSshy no, the new toolchains are the nightly snapshot builds

**Comment by compnerd** [comment]
@stevapple btw, assuming that testing goes well with #295, once that is merged, the `swiftc` invocation will change to:

```
swiftc -sdk %SDKROOT% -libc MD -O -emit-executable -o executable.exe executable.swift
swiftc -sdk %SDKROOT% -libc MD -O -parse-as-library -emit-library -o library.dll library.swift
```

Note that `SDKROOT` is actually setup by the installer (at least for the time being), so that should pretty much just work in any x64 Developer Command Prompt out of the box.

**Comment by klcantrell** [comment]
Hey guys, excited to see Swift come to Windows, thanks for your work.

I'm not sure what I'm missing, but given the following setup:

1. Download and run installer.exe from the latest build snapshots ([artifacts I downloaded from](https://compnerd.visualstudio.com/swift-build/_build/results?buildId=38642&view=artifacts&type=publishedArtifacts)).
2. Setup requirements/installation per the [Getting Started Windows](https://github.com/compnerd/swift-build/blob/master/docs/GettingStartedWindows.md) doc.

I am able to successful initialize a project with `swift package init --type=executable`. However, when I `swift run` in the project, it just hangs there forever with no `Hello, world!` output. `swift build` also just hangs there forever.

Any advice would be much appreciated, thanks!

**Comment by compnerd** [comment]
That is interesting; are you running that from the "X64 Developer Command Prompt"?  Can you run `swift run -v` and see what it is trying to run that is hanging?

**Comment by klcantrell** [comment]
I'm running from the "X64 Native Tools Command Prompt for VS 2019" (I think that's the same one you mentioned, although the name is different). The thing it's trying to run but stuck on is `C:\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin\swiftc.exe -print-target-info`.

**Comment by compnerd** [comment]
I just ran that specific build on a clean install, it seems to work fine for me.  It did however require an extra return to refresh the console.

**Comment by klcantrell** [comment]
Okay, thanks for looking into it! I tried the extra return but still no luck on my end. I may try reinstalling some things to see if that helps.

I'm actually quite stoked that I can just get `swiftc` working. I know that SPM for Windows is still a work in progress, but looking forward to when it's more accessible.

**Comment by compnerd** [comment]
@klcantrell one more thing - did you install 5.3 and then upgrade to the latest snapshot?  If so, you probably skipped over an important step: the support files need to be refreshed anytime you update the swift or visual studio installation.

**Comment by klcantrell** [comment]
That must have been it! Yeah, I did install 5.3 first then update. Interestingly, when re-running the following to update the support files:
```
copy %SDKROOT%\usr\share\ucrt.modulemap "%UniversalCRTSdkDir%\Include\%UCRTVersion%\ucrt\module.modulemap"
copy %SDKROOT%\usr\share\visualc.modulemap "%VCToolsInstallDir%\include\module.modulemap"
copy %SDKROOT%\usr\share\visualc.apinotes "%VCToolsInstallDir%\include\visualc.apinotes"
copy %SDKROOT%\usr\share\winsdk.modulemap "%UniversalCRTSdkDir%\Include\%UCRTVersion%\um\module.modulemap"
```
It prompted me with every line to overwrite, so it seemed the support files were still there. I went ahead and overwrote, then retried `swift run` and it worked!

Another interesting thing is that I can't seem to do `swift run` from a non-Administrator prompt. It seems to be required for me to run as an administrator otherwise I get the following:

```
error: manifest parse error(s):
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\visualc.apinotes:1:1: error: not a mapping
//===--- ucrt.modulemap ---------------------------------------------------===//
^
<module-includes>:1:10: note: in file included from <module-includes>:1:
#include "AssertionReporting.h"
         ^
C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\shims/AssertionReporting.h:16:10: note: in file included from C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\shims/AssertionReporting.h:16:
#include "SwiftStdint.h"
         ^
C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\shims/SwiftStdint.h:28:10: error: could not build module 'visualc'
#include <stdint.h>
         ^
C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\shims/SwiftStddef.h:25:10: note: while building module 'ucrt' imported from C:\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\shims/SwiftStddef.h:25:
#include <stddef.h>
         ^
<module-includes>:1:10: note: in file included from <module-includes>:1:
#include "complex.h"
         ^
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt/complex.h:12:10: note: in file included from C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt/complex.h:12:
#include <corecrt.h>
         ^
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt\corecrt.h:10:10: error: could not build module 'visualc'
#include <vcruntime.h>
         ^
<unknown>:0: error: could not build C module 'SwiftShims'
```

**Comment by compnerd** [comment]
That is a copy-paste issue.  I think that `visualc.apinotes` has the wrong content.  You may have to also clear out `.build`

**Comment by klcantrell** [comment]
Clearing out `.build` and re-running `swift build` seems to have done the trick! Thanks for your help @compnerd! 💯

**Comment by compnerd** [comment]
I don't think that there is much left to do here; the `swiftc` usage is down to:
```
swiftc -sdk %SDKROOT% -libc MD -emit-executable -o executable.exe executable.swift
swiftc -sdk %SDKROOT% -libc MD -parse-as-library -emit-library -o library.dll library.swift
```
and with SPM, it is just down to `swift build`.
